src/keri/mod.rs

```diff
@@ -198,6 +365,46 @@ impl<'d, K: KeyManager> Keri<'d, K> {
         Ok(signed_rcp)
     }
 
+    /// Create `SignedNontransferableReceipt` for given `EventMessage`
+    /// This will actually process and generate receipt if we are added as witness
+    /// Generated receipt will be stored into `ntp` DB table under sender's identifier
+    /// Ignore and return `Error::SemanticError` with description why no receipt returned
+    ///
+    /// # Parameters
+    /// * `message` - `EventMessage` we are to process
+    ///
+    pub fn make_ntr(&self, message: EventMessage) -> Result<SignedNontransferableReceipt, Error> {
+        let our_bp = match &self.prefix {
+            IdentifierPrefix::Basic(prefix) => prefix,
+            _ => return Err(Error::SemanticError("we are not a witness - our prefix is not Basic".into()))
+        };
+        match &message.event.event_data {
+            // ICP requires check if we are in initial witnesses only
+            EventData::Icp(evt) => {
+                if !evt.witness_config.initial_witnesses.contains(&our_bp) {
+                    return Err(Error::SemanticError("we are not in a witness list.".into()));
+                }
+                self.generate_ntr(message)
+            },
+            EventData::Rot(evt) => {
+                if !evt.witness_config.prune.contains(&our_bp) {
+                    if evt.witness_config.graft.contains(&our_bp) {
+                        // FIXME: logic for already witnessed identifier required
+                        self.generate_ntr(message)
+                    } else {
+                        Err(Error::SemanticError("event does not change our status as a witness".into()))
+                    }
+                } else if evt.witness_config.prune.contains(&our_bp) {
+                    self.processor.db.remove_receipts_nt(&message.event.prefix)?;
+                    Err(Error::SemanticError("we were removed. no receipt to generate".into()))
+                } else {
+                    Err(Error::SemanticError("event without witness modifications".into()))
```

Comment:
Is it even possible to get here?

Reply:
yup, this one is badly formed. should rotate it with only one `.contains()` check. will fix...

Reply:
nope, double checked the code and the logic is next:
if we are not pruned, but grafted -> Ok(ntr)
else if we are pruned ->  Remove(ntr), nothing to gen
else (nothing about us OR both prune and graft) -> We do nothing in regards to us as a witness

Reply:
the only issue here is that if in rot event we are both pruned and grafted - prune prevails and we are no longer considering ourselves as witnesses. but in that case is it a malformed rot event or what should be the logic? that's why spec is important to have =)

Reply:
But we will never get to _nothing about us_ branch, because we will get into _we are not pruned_ branch first :wink: I think we don't need the last else.

Reply:
we won't, but, as it is a returning if/else, it cannot go without else. and to do it with else, but without else if we'll have to do double check of `.contains()` in both if and else statements. consider this a syntactic sugar to make compiler happy =)

Reply:
How about match instead of ifs? Something like this: 
```rust
 match (evt.witness_config.prune.contains(&our_bp), evt.witness_config.graft.contains(&our_bp)) {
     (true, _) => {
         self.processor.db.remove_receipts_nt(&message.event.prefix)?;
         Err(Error::SemanticError("we were removed. no receipt to generate".into()))
     },
     (false, true) =>  self.generate_ntr(message),
     (false, false) =>  Err(Error::SemanticError("event does not change our status as a witness".into())),
}
```

Reply:
that's same burden of double checking each time =)
in my approach double check happens only in case of !prune...
it might seem minor, but it's (N(of prune) + N(of graft) on each rotation or inception validation) * N(messages parsed),
which for witnesses can be a lot...